File: frontend/src/api/saved.ts
import type { CohortFilter } from './cohorts'

// ── Saved cohorts ────────────────────────────────────────────────────────────

export interface SavedCohortIn {
  name: string
  description?: string | null
  filters: CohortFilter
}

export interface SavedCohort {
  cohort_id: string
  name: string
  description: string | null
  filter_json: string
  member_count: number
  user_email: string | null
  created_at: string
}

export interface SavedCohortDetail extends SavedCohort {
  members: string[]
}

export interface SavedCohortList {
  items: SavedCohort[]
}

// ── Feature runs ─────────────────────────────────────────────────────────────

export interface SaveFeatureRunIn {
  name: string
  description?: string | null
  usubjid: string
  day_min: number
  day_max: number
  window_day_start?: number | null
  window_day_end?: number | null
  feature_ids: string[]
}

export interface FeatureRunSummary {
  run_id: string
  name: string | null
  description: string | null
  usubjid: string
  study_day_start: number | null
  study_day_end: number | null
  n_features: number
  n_rows: number
  user_email: string | null
  created_at: string
}

export interface FeatureRunList {
  items: FeatureRunSummary[]
}

export interface FeatureRunValue {
  feature_id: string
  value_key: string
  value: number | null
}

export interface FeatureRunDetail extends FeatureRunSummary {
  values: FeatureRunValue[]
  params_json: string | null
  gcs_paths: string[]
}

// ── Feature sets ─────────────────────────────────────────────────────────────

export interface FeatureSetIn {
  name: string
  description?: string | null
  feature_ids: string[]
  params_json?: string | null
}

export interface FeatureSet {
  feature_set_id: string
  name: string
  description: string | null
  feature_ids: string[]
  params_json: string | null
  user_email: string | null
  created_at: string
}

export interface FeatureSetList {
  items: FeatureSet[]
}

File: backend/app/schemas/saved.py
from pydantic import BaseModel, Field

from .cohorts import CohortFilter


# ── Cohorts ──────────────────────────────────────────────────────────────────


class SavedCohortIn(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    description: str | None = None
    filters: CohortFilter = Field(default_factory=CohortFilter)


class SavedCohort(BaseModel):
    cohort_id: str
    name: str
    description: str | None = None
    filter_json: str
    member_count: int
    user_email: str | None = None
    created_at: str


class SavedCohortDetail(SavedCohort):
    members: list[str] = Field(default_factory=list)


class SavedCohortList(BaseModel):
    items: list[SavedCohort]


# ── Feature runs ─────────────────────────────────────────────────────────────


class SaveFeatureRunIn(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    description: str | None = None
    usubjid: str
    day_min: int
    day_max: int
    window_day_start: float | None = None
    window_day_end: float | None = None
    feature_ids: list[str] = Field(min_length=1, max_length=50)


class FeatureRunSummary(BaseModel):
    run_id: str
    name: str | None = None
    description: str | None = None
    usubjid: str
    study_day_start: float | None = None
    study_day_end: float | None = None
    n_features: int
    n_rows: int
    user_email: str | None = None
    created_at: str


class FeatureRunList(BaseModel):
    items: list[FeatureRunSummary]


class FeatureRunValue(BaseModel):
    feature_id: str
    value_key: str
    value: float | None = None


class FeatureRunDetail(FeatureRunSummary):
    values: list[FeatureRunValue] = Field(default_factory=list)
    params_json: str | None = None
    gcs_paths: list[str] = Field(default_factory=list)


# ── Feature sets ─────────────────────────────────────────────────────────────


class FeatureSetIn(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    description: str | None = None
    feature_ids: list[str] = Field(min_length=1, max_length=50)
    params_json: str | None = None


class FeatureSet(BaseModel):
    feature_set_id: str
    name: str
    description: str | None = None
    feature_ids: list[str]
    params_json: str | None = None
    user_email: str | None = None
    created_at: str


class FeatureSetList(BaseModel):
    items: list[FeatureSet]
